runtime/registry.go

```diff
@@ -79,11 +72,148 @@ func (r *Runtime) CreateInstance(ctx context.Context, inst *drivers.Instance) er
+		for _, entry := range entries {
+			if entry.Type == drivers.ObjectTypeTable {
+				table := entry.GetTable()
+				if table.Managed {
+					dropErrors = append(dropErrors, svc.Olap.Exec(ctx, &drivers.Statement{Query: fmt.Sprintf("DROP TABLE %s", table.Name)}))
+				}
+			}
+		}
+		// drop rill schema which deletes embedded catalog as well
+		dropErrors = append(dropErrors, svc.Olap.Exec(ctx, &drivers.Statement{Query: "DROP schema rill CASCADE"}))
+	}
+	// delete instance related data if catalog is not embedded
+	if !inst.EmbedCatalog {
+		dropErrors = append(dropErrors, svc.Catalog.DeleteInstanceEntries(ctx, instanceID))
+	}
+
+	// evict caches
+	r.evictCaches(ctx, inst)
+
+	// delete instance
+	err4 := r.Registry().DeleteInstance(ctx, instanceID)
+	return util.ReturnFirstErr(util.ReturnFirstErr(err1, err4), util.ReturnFirstErr(dropErrors...))
+}
+
+// EditInstance edits exisiting instance.
+// Confirming to put api specs, it is expected to send entire existing instance data.
+// The API compares and only evicts caches drivers or dsn is changed.
+// This is done to ensure that db handlers are not unnecessarily closed
+func (r *Runtime) EditInstance(ctx context.Context, inst *drivers.Instance) error {
+	olderInstance, err := r.Registry().FindInstance(ctx, inst.ID)
+	if err != nil {
+		return err
+	}
+
+	// 1. changes in olap driver or olap dsn
+	var olapConn drivers.Connection
+	olapChanged := olderInstance.OLAPDriver != inst.OLAPDriver || olderInstance.OLAPDSN != inst.OLAPDSN
+	if olapChanged {
+		// Check OLAP connection
+		olap, _, err := r.checkOlapConnection(inst)
+		if err != nil {
+			return err
+		}
+
+		// Prepare connections for use
+		err = olap.Migrate(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to prepare instance: %w", err)
+		}
+		olapConn = olap
+	}
+
+	// 2. embedCatalog disabled previously but enabled now
+	if inst.EmbedCatalog {
+		if olapConn == nil {
+			// getting exisiting connection
+			olapConn, err = r.connCache.get(ctx, inst.ID, inst.OLAPDriver, inst.OLAPDSN)
+			if err != nil {
+				return err
+			}
+		}
+		_, ok := olapConn.CatalogStore()
+		if !ok {
+			return errors.New("driver does not support embedded catalogs")
+		}
+	}
+
+	// 3. changes in repo driver or repo dsn
+	repoChanged := inst.RepoDriver != olderInstance.RepoDriver || inst.RepoDSN != olderInstance.RepoDSN
+	if repoChanged {
+		// Check repo connection
+		repo, _, err := r.checkRepoConnection(inst)
+		if err != nil {
+			return err
+		}
+
+		// Prepare connections for use
+		err = repo.Migrate(ctx)
+		if err != nil {
+			return fmt.Errorf("failed to prepare instance: %w", err)
+		}
+	}
+
+	// evict caches if connections need to be updated
+	if olapChanged || repoChanged {
+		r.evictCaches(ctx, olderInstance)
+	}
+
+	// update env variables
+	if inst.Env == nil {
+		inst.Env = make(map[string]string)
+	}
+	inst.Env["allow_host_credentials"] = strconv.FormatBool(r.opts.AllowHostCredentials)
```

Comment:
This doesn't actually update the env?

Reply:
Sorry didn't understand this. In case env's is updated this needs to be added as well. If not, this will just set the older value.

Reply:
Sorry, my fault, of course this is the new env. You can disregard my comment